pkg/planner/core/logical_plan_builder.go

```diff
@@ -4868,10 +4872,39 @@ func (b *PlanBuilder) tryBuildCTE(ctx context.Context, tn *ast.TableName, asName
 			prevSchema := cte.seedLP.Schema().Clone()
 			lp.SetSchema(getResultCTESchema(cte.seedLP.Schema(), b.ctx.GetSessionVars()))
 
-			if cte.recurLP != nil && cte.isInline {
-				b.ctx.GetSessionVars().StmtCtx.AppendWarning(
-					ErrInternal.GenWithStack("Recursive CTE can not be inlined."))
+			// If current CTE query contain another CTE which 'containAggOrWindow' is true, current CTE 'containAggOrWindow' will be true
+			if b.buildingCTE {
+				b.outerCTEs[len(b.outerCTEs)-1].containAggOrWindow = cte.containAggOrWindow || b.outerCTEs[len(b.outerCTEs)-1].containAggOrWindow
+			}
+
+			// Combine the declaration of CTE and the use of CTE to jointly determine **whether a CTE can be inlined**
+			/*
+			   There are some cases that CTE must be not inlined.
+			   1. CTE is recursive CTE.
+			   2. CTE contains agg or window and it is referenced by recursive part of CTE.
+			   3. Consumer count of CTE is more than one.
+			   If 1 or 2 conditions are met, CTE cannot be inlined.
+			   But if query is hint by 'merge()' or session variable "tidb_opt_force_inline_cte",
+			     CTE will still not be inlined but a warning will be recorded "Hint or session variables are invalid"
+			   If 3 condition is met, CTE can be inlined by hint and session variables.
+			*/
+			if cte.recurLP != nil {
+				if cte.forceInlineByHintOrVar {
+					b.ctx.GetSessionVars().StmtCtx.AppendWarning(
+						ErrInternal.GenWithStack("Recursive CTE %s can not be inlined by merge() or tidb_opt_force_inline_cte.", tn.Name))
+				}
+			} else if cte.containAggOrWindow && b.buildingRecursivePartForCTE {
+				if cte.forceInlineByHintOrVar {
+					b.ctx.GetSessionVars().StmtCtx.AppendWarning(ErrCTERecursiveForbidsAggregation.FastGenByArgs(tn.Name))
+				}
+			} else if cte.consumerCount > 1 {
+				if cte.forceInlineByHintOrVar {
+					cte.isInline = true
+				}
+			} else {
+				cte.isInline = true
```

Comment:
Cool, this part code is much more clear than before, but could you create a separate function for checking CTE inline instead of just putting it in the big function `buildSelect`...

Reply:
Very good suggestion ~ Done